File: 0274_h_index/h_index.java
// Time: O(n)
// Space: O(n)
class Solution {
    public int hIndex(int[] citations) {
        int len = citations.length;
        int[] bucket = new int[len + 1];
        for (int c : citations)
            if (c > len)
                bucket[len]++;
            else
                bucket[c]++;
        int res = 0;
        for (int i = len; i >= 0; i--) {
            res += bucket[i];
            if (res >= i)
                return i;
        }
        return 0;
    }
}

File: 0274_h_index/h_index.py
from typing import List


def h_index(citations: List[int]) -> int:
    """
    Time: O(n)
    Space: O(n)
    """
    size = len(citations)
    bucket = [0] * (size + 1)
    for c in citations:
        if c >= size:
            bucket[size] += 1
        else:
            bucket[c] += 1
    i = size
    res = 0
    while i >= 0:
        res += bucket[i]
        if res >= i:
            return i
        i -= 1
    return 0
